doc/catalyst-cli/catalyst-cli.rst

```diff
@@ -0,0 +1,237 @@
+applies two quantum-optimization passes:
+
+#. ``remove-chained-self-inverse``, which removes any operations that are applied next to their
+   (self-)inverses or adjoint, in this case the two adjacent Hadamard gates.
+#. ``merge-rotations``, which combines rotation gates of the same type that act sequentially, in
+   this case the two RX gates the become adjacent after the two Hadamard gates have been removed by
+   the ``remove-chained-self-inverse`` pass.
+
+To define the pass pipeline, we must supply the name of the function to which each pass applies
+using the ``func-name`` argument. To apply these two passes to our ``my_circuit`` function, we can
+do so as follows:
+
+.. code-block::
+
+    pipe(remove-chained-self-inverse{func-name=my_circuit};merge-rotations{func-name=my_circuit})
+
+Finally, we'll use the option ``--mlir-print-ir-after-all`` to print the resulting MLIR after each
+pass that is applied, and the ``-o`` option to set the name of the output file:
+
+.. code-block::
+
+    catalyst-cli my_circuit.mlir \
+        --tool=opt \
+        --catalyst-pipeline="pipe(remove-chained-self-inverse{func-name=my_circuit};merge-rotations{func-name=my_circuit})" \
+        --mlir-print-ir-after-all \
+        -o my_circuit-llvm.mlir
+
+Running this command will output the following intermediate IR to the console:
+
+.. code-block:: mlir
+
+    // -----// IR Dump After RemoveChainedSelfInversePass (remove-chained-self-inverse) //----- //
+    module {
+      func.func @my_circuit(%arg0: !quantum.bit, %arg1: f64) -> !quantum.bit {
+        %out_qubits = quantum.custom "RX"(%arg1) %arg0 : !quantum.bit
+        %out_qubits_0 = quantum.custom "RX"(%arg1) %out_qubits : !quantum.bit
+        return %out_qubits_0 : !quantum.bit
+      }
+    }
+
+
+    // -----// IR Dump After MergeRotationsPass (merge-rotations) //----- //
+    module {
+      func.func @my_circuit(%arg0: !quantum.bit, %arg1: f64) -> !quantum.bit {
+        %0 = arith.addf %arg1, %arg1 : f64
+        %out_qubits = quantum.custom "RX"(%0) %arg0 : !quantum.bit
+        return %out_qubits : !quantum.bit
+      }
+    }
+
+and produce a new file ``my_circuit-llvm.mlir`` containing the resulting module in the LLVM MLIR
+dialect:
+
+.. code-block:: mlir
+
+    module {
+      func.func @my_circuit(%arg0: !quantum.bit, %arg1: f64) -> !quantum.bit {
+        %0 = arith.addf %arg1, %arg1 : f64
+        %out_qubits = quantum.custom "RX"(%0) %arg0 : !quantum.bit
+        return %out_qubits : !quantum.bit
+      }
+    }
+
+We can see in the intermediate IR after the ``remove-chained-self-inverse`` pass that the two
+adjacent Hadamard gates were removed and that the two RX gates were merged into one after the
+``merge-rotations`` pass, with the input angle to the single RX gate being the sum of the two input
+angles to the original two gates. The result in ``my_circuit-llvm.mlir`` contains the final,
+optimized MLIR.
```

Comment:
Can you add a line like "for a list of Catalyst's available compiler passes, see --help"?

Reply:
(We should really have that list in the documentation itself... but we've been procrastinating for a while now on this I guess :sweat: )

Reply:
(We do have links to source code of the passes in the documentation https://docs.pennylane.ai/projects/catalyst/en/stable/dev/transforms.html#catalyst-s-transformation-library, so maybe we can link to that here as well)

Reply:
I'm not sure how helpful it would be to a user to run `--help` to see the list of available compiler passes, since they're listed as options along with all the other CLI options, e.g.

```
--remove-chained-self-inverse   -   Perform removal of chained operations that are self inverses.
  --func-name=<string>          - The name of the function (or qnode) to run the quantum circuit transformation pass on.
```

It wouldn't be clear to me reading through the ~800 lines of options which options correspond to Catalyst passes!

I think it makes more sense to link to the "Catalyst's Transformation Library" documentation for now, but we _should_ eventually list out the passes explicitly there.

Reply:
Added in 8e3eaca0

I added the cross-reference to the "Catalyst's Transformation Library" docs, and as an aside, also said that the available passes are listed in the `--help` message.

Reply:
I agree that `--help` dump is not ideal, but since it's the only way to get a list of available passes at command line, I still think we should have it.

Reply:
Yes I agree :slightly_smiling_face:

The updated version is live at https://xanaduai-pennylane--1249.com.readthedocs.build/projects/catalyst/en/1249/catalyst-cli/catalyst-cli.html